gtsfm/runner/gtsfm_runner_base.py

```diff
@@ -199,14 +205,25 @@ def run(self) -> GtsfmData:
                 os.path.join(self.parsed_args.output_root, "gtsfm", "configs", self.parsed_args.cluster_config)
             )["workers"]
             scheduler = workers[0]
-            cluster = SSHCluster(
-                [scheduler] + workers,
-                scheduler_options={"dashboard_address": self.parsed_args.dashboard_port},
-                worker_options={
-                    "n_workers": self.parsed_args.num_workers,
-                    "nthreads": self.parsed_args.threads_per_worker,
-                },
-            )
+            connected = False
+            retry_count = 0
+            while retry_count < self.parsed_args.num_retry_cluster_connection and not connected:
+                logger.info(f"Connecting to the cluster: attempt {retry_count + 1}")
+                logger.info(f"Using {scheduler} as scheduler")
+                logger.info(f"Using {workers} as workers")
+                try:
+                    cluster = SSHCluster(
+                        [scheduler] + workers,
+                        scheduler_options={"dashboard_address": self.parsed_args.dashboard_port},
+                        worker_options={
+                            "n_workers": self.parsed_args.num_workers,
+                            "nthreads": self.parsed_args.threads_per_worker,
+                        },
+                    )
+                    connected = True
+                except Exception as e:
+                    logger.info(f"Worker failed to start: {str(e)}")
+                    retry_count += 1
```

Comment:
Could you move this entire logic to a separate `setup_ssh_cluster_with_retries` function?

Reply:
@stepanyanhayk lgtm except I agree with Akshay, this code would be cleaner if we decomposed it into a helper

Reply:
Done.